File: packages/adapters/src/ordering.ts
/**
 * Unicode code-point ordering. The contract orders capabilities, hosts,
 * environment names, provider metrics, usage resources and provider-safe detail
 * fields by code point so that two runtimes in two languages produce the same
 * canonical document.
 */

export function compareUnicodeCodePoints(left: string, right: string): number {
  const leftPoints = Array.from(left, (value) => value.codePointAt(0) ?? 0);
  const rightPoints = Array.from(right, (value) => value.codePointAt(0) ?? 0);
  const common = Math.min(leftPoints.length, rightPoints.length);
  for (let index = 0; index < common; index += 1) {
    const a = leftPoints[index] as number;
    const b = rightPoints[index] as number;
    if (a !== b) return a - b;
  }
  return leftPoints.length - rightPoints.length;
}

/** Strictly ascending: equal neighbours are a duplicate, not an order. */
export function isSortedByCodePoint(values: readonly string[]): boolean {
  for (let index = 1; index < values.length; index += 1) {
    const previous = values[index - 1] as string;
    const current = values[index] as string;
    if (compareUnicodeCodePoints(previous, current) >= 0) return false;
  }
  return true;
}

export function sortByCodePoint(values: readonly string[]): string[] {
  return [...values].sort(compareUnicodeCodePoints);
}

File: python/src/graph_engineering/adapters/ordering.py
"""Unicode code-point ordering.

The contract orders capabilities, hosts, environment names, provider metrics,
usage resources and provider-safe detail fields by code point so that two
runtimes in two languages produce the same canonical document.  Python compares
``str`` by code point already, so this module is the naming of that fact rather
than an algorithm.
"""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from itertools import pairwise


def compare_unicode_code_points(left: str, right: str) -> int:
    if left < right:
        return -1
    if left > right:
        return 1
    return 0


def is_sorted_by_code_point(values: Sequence[str]) -> bool:
    """Strictly ascending: equal neighbours are a duplicate, not an order."""
    return all(previous < current for previous, current in pairwise(values))


def sort_by_code_point(values: Iterable[str]) -> tuple[str, ...]:
    return tuple(sorted(values))
